aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt

```diff
@@ -1612,4 +1626,136 @@ internal class RealAWSCognitoAuthPlugin(
+                            "The user is currently federated to identity pool. You " +
+                                "must call clearFederationToIdentityPool to clear credentials."
+                        )
+                    )
+                }
+                (
+                    authNState is AuthenticationState.SignedOut ||
+                        authNState is AuthenticationState.Error ||
+                        authNState is AuthenticationState.NotConfigured
+                    ) || (
+                    authZState is AuthorizationState.Configured ||
+                        authZState is AuthorizationState.SessionEstablished ||
+                        authZState is AuthorizationState.Error
+                    ) -> {
+
+                    _federateToIdentityPool(authProvider, providerToken, options, onSuccess, onError)
+                }
+                else -> onError.accept(
+                    AuthException.InvalidStateException("Federation could not be completed.")
+                )
+            }
+        }
+    }
+
+    private fun _federateToIdentityPool(
+        authProvider: AuthProvider,
+        providerToken: String,
+        options: FederateToIdentityPoolOptions?,
+        onSuccess: Consumer<FederateToIdentityPoolResult>,
+        onError: Consumer<AuthException>
+    ) {
+        var token: StateChangeListenerToken? = null
+        token = authStateMachine.listen(
+            { authState ->
+                val authNState = authState.authNState
+                val authZState = authState.authZState
+                when {
+                    authNState is AuthenticationState.FederatedToIdentityPool
+                        && authZState is AuthorizationState.SessionEstablished -> {
+                        token?.let(authStateMachine::cancel)
+                        val credential = authZState.amplifyCredential as? AmplifyCredential.IdentityPoolFederated
+                        val identityId = credential?.identityId
+                        val awsCredentials = credential?.credentials
+                        if (identityId != null && awsCredentials != null) {
+                            val result = FederateToIdentityPoolResult(
+                                credentials = awsCredentials,
+                                identityId = identityId
+                            )
+                            onSuccess.accept(result)
+                        } else {
+                            onError.accept(
+                                AuthException.UnknownException(
+                                    "Could not start federation to Identity Pool",
+                                    "An unclassified error prevented this operation."
+                                )
+                            )
+                        }
+                    }
+                    authNState is AuthenticationState.Error && authZState is AuthorizationState.Error -> {
```

Comment:
Since there are no actions being called from authentication state during federation flow, don't know why AuthenticationState would ever be in Error state.

Reply:
It can happen here:
https://github.com/aws-amplify/amplify-android/blob/04e32971e6d06b79771b6eb1be754a081c117b2c/aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/states/AuthenticationState.kt#L139

This was direct follow from iOS SM